File: client/src/app/core/services/room.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { WebSocketConnectionService } from './websocket-connection.service';
import { NGXLogger } from 'ngx-logger';

@Injectable({
  providedIn: 'root',
})
export class RoomService {
  public rooms$ = new BehaviorSubject<string[]>([]);
  public members$ = new BehaviorSubject<string[]>([]);
  public currentRoom = 'main';

  constructor(
    private wsService: WebSocketConnectionService,
    private logger: NGXLogger
  ) {
    this.wsService.systemMessages$.subscribe((message) => {
      this.handleSystemMessage(message);
    });
  }

  public listRooms(): void {
    this.logger.info('listRooms', 'Listing rooms');
    this.wsService.send('[UserCommand] /list');
  }

  public joinRoom(room: string): void {
    if (room !== this.currentRoom) {
      this.wsService.send(`[UserCommand] /join ${room}`);
      this.currentRoom = room;
    } else {
      this.logger.warn('joinRoom', `Already in room: ${room}`);
    }
  }

  private handleSystemMessage(message: string): void {
    if (message.includes('[SystemRooms]')) {
      const matchRooms = message.match(/\[SystemRooms]\s*(.*?)$/);
      if (matchRooms) {
        const rooms = matchRooms[1].split(',').map((room: string) => room.trim());
        this.rooms$.next(rooms);
      } else {
        this.logger.warn('handleSystemMessage', `No rooms found in message: ${message}`);
      }
    } else if (message.includes('[SystemMembers]')) {
      const matchMembers = message.match(/\[SystemMembers]\s*(.*?)$/);
      if (matchMembers) {
        const members = matchMembers[1].split(',').map((member: string) => member.trim());
        this.members$.next(members);
      } else {
        this.logger.warn('handleSystemMessage', `No members found in message: ${message}`);
      }
    } else if (message.includes('[SystemJoin]')) {
      const matchJoin = message.match(/^(.*?)\s*\[SystemJoin]\s*(.*?)$/);
      if (matchJoin) {
        this.logger.info('handleSystemMessage', `User joined room ${matchJoin[2]}`);
        this.currentRoom = matchJoin[2];
      } else {
        this.logger.warn('handleSystemMessage', `No room to join found in message: ${message}`);
      }
    }
  }
}

File: client/ios/PastePoint/Core/Services/RoomService.swift
//
//  Copyright © 2026 PastePoint. All rights reserved.
//  SPDX-License-Identifier: GPL-3.0-only
//

import Combine
import Foundation
import Logging

@MainActor
final class RoomService: ObservableObject {
    private let logger = Logger(label: "Room")

    @Published var rooms: [String] = []
    @Published var members: [String] = []
    @Published var currentRoom: String = ""

    private var cancellables = Set<AnyCancellable>()
    private let wsService: WebSocketConnectionService

    init(wsService: WebSocketConnectionService) {
        self.wsService = wsService

        wsService.systemMessage
            .receive(on: DispatchQueue.main)
            .sink { [weak self] message in
                self?.handleSystemMessage(message)
            }
            .store(in: &cancellables)

        wsService.didConnect
            .receive(on: DispatchQueue.main)
            .sink { [weak self] _ in
                Task { await self?.listRooms() }
            }
            .store(in: &cancellables)
    }

    func listRooms() async {
        await wsService.send("[UserCommand] /list")
    }

    func joinOrCreateRoom(_ room: String) async {
        guard !room.isEmpty else {
            logger.warning("joinOrCreateRoom: room name is empty — skipped")
            return
        }
        guard room != currentRoom else {
            logger.debug("joinOrCreateRoom: already in room '\(room)' — skipped")
            return
        }
        logger.info("Joining room: \(room)")
        await wsService.send("[UserCommand] /join \(room)")
        currentRoom = room
        await listRooms()
    }

    private func handleSystemMessage(_ message: String) {
        if message.contains("[SystemRooms]") {
            guard let range = message.range(of: "\\[SystemRooms]\\s*(.*)$", options: .regularExpression) else {
                logger.warning("handleSystemMessage: failed to parse [SystemRooms] message: \(message)")
                return
            }
            let rest = String(message[range])
            let prefix = "[SystemRooms] "
            let list = rest.hasPrefix(prefix) ? String(rest.dropFirst(prefix.count)) : rest
            rooms = list.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }
            logger.debug("Rooms updated: \(rooms)")
        } else if message.contains("[SystemMembers]") {
            guard let range = message.range(of: "\\[SystemMembers]\\s*(.*)$", options: .regularExpression) else {
                logger.warning("handleSystemMessage: failed to parse [SystemMembers] message: \(message)")
                return
            }
            let rest = String(message[range])
            let prefix = "[SystemMembers] "
            let list = rest.hasPrefix(prefix) ? String(rest.dropFirst(prefix.count)) : rest
            members = list.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }
            logger.debug("Members updated: \(members)")
        } else if message.contains("[SystemJoin]") {
            guard let range = message.range(of: "\\[SystemJoin]\\s*(\\S+)\\s*$", options: .regularExpression) else {
                logger.warning("handleSystemMessage: failed to parse [SystemJoin] message: \(message)")
                return
            }
            let parts = String(message[range]).split(separator: " ")
            guard let last = parts.last else {
                logger.warning("handleSystemMessage: [SystemJoin] had no room name in: \(message)")
                return
            }
            currentRoom = String(last)
            logger.info("Joined room: \(currentRoom)")
            Task { await self.listRooms() }
        }
    }
}
